### PR TITLE
Bulk subscribe: Sending cloudEvent for non-rawPayload 

### PR DESCRIPTION
**Pull request by DeepanshuA**
# Description

As explained in issue #5369, it was discussed and decided to send a cloudEvent instead of bytes directly for non-rawPayload event, for grpc.
This change addresses that.

## Issue reference

Please reference the issue this PR will close: #5369 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Review by mukundansundar — CHANGES_REQUESTED** [review]
Please add extensions to this proto as well similar to https://github.com/dapr/dapr/pull/5516/files

File: dapr/proto/runtime/v1/appcallback.proto [hunk 1]
```diff
@@ -114,19 +114,43 @@ message TopicEventResponse {
...
+  string type = 3;
+
+  // The version of the CloudEvents specification. 
+  string spec_version = 4;
+
+  // The content type of data value.
+  string data_content_type = 5;
+
+  // The content of the event.
+  bytes data = 6;
```
Comment (berndverst) [inline dapr/proto/runtime/v1/appcallback.proto:135]: Please add

```suggestion
  bytes data = 6;
  
  // Custom attributes which includes cloud event extensions.
  google.protobuf.Struct extensions = 7;
```

See:
https://github.com/dapr/dapr/pull/5516

**Review by berndverst — CHANGES_REQUESTED** [review]
Please replicate the code used for custom attributes (cloud event extensions) for the regular gRPC publish path. This was just merged https://github.com/dapr/dapr/pull/5516.

The entire cloud event represents a map[string]interface{}. All properties minus the well known cloud event fields (id, source, type, spec version, data content type, data field / data field base64) should be removed from the cloud event, and the remainder should be passed through as a struct on the proto.

File: dapr/proto/runtime/v1/appcallback.proto (hunk 1 above)
Comment (DeepanshuA) [reply dapr/proto/runtime/v1/appcallback.proto:139]: done

**Comment by DeepanshuA** [comment]
> Please replicate the code used for custom attributes (cloud event extensions) for the regular gRPC publish path. This was just merged #5516.
> 
> The entire cloud event represents a map[string]interface{}. All properties minus the well known cloud event fields (id, source, type, spec version, data content type, data field / data field base64) should be removed from the cloud event, and the remainder should be passed through as a struct on the proto.

I have incorporated suggested change, please re-review @berndverst @mukundansundar

File: pkg/runtime/bulk_subscriber.go [hunk 2]
```diff
@@ -433,20 +433,80 @@ func (a *DaprRuntime) publishBulkMessageHTTP(ctx context.Context, msg *pubsubBul
...
+
 // publishBulkMessageGRPC publishes bulk message to a subscriber using gRPC and takes care of corresponding responses.
 func (a *DaprRuntime) publishBulkMessageGRPC(ctx context.Context, msg *pubsubBulkSubscribedMessage,
 	bulkResponses *[]pubsub.BulkSubscribeResponseEntry, entryIdIndexMap map[string]int, bulkSubDiag *bulkSubIngressDiagnostics, //nolint:stylecheck
 ) error {
 	items := make([]*runtimev1pb.TopicEventBulkRequestEntry, len(msg.entries))
+	rawPayload := true
+	if msg.cloudEvents[0] != nil {
+		rawPayload = false
+	}
```
Comment (mukundansundar) [inline pkg/runtime/bulk_subscriber.go:499]: can you base this off of the actual metadata input by the user instead of basing it off of cloudEvents list?

**Review by mukundansundar — COMMENTED** [review]
Please add UTs for the case where message is got as TopicEventCERequest. 
Please add UTs for extensions in bulk subscribe.

**Comment by DeepanshuA** [comment]
> Please add UTs for the case where message is got as TopicEventCERequest. Please add UTs for extensions in bulk subscribe.

TopicEventCERequest is already being used in GRPC UTs implicitly. I have added verification for data as well in a UT, which takes value from CloudEvent proto i.e. TopicEventCERequest and verifies against supplied expected.
Added extension as well in verification in UT.

File: pkg/runtime/bulk_subscriber.go (hunk 2 above)
Comment (DeepanshuA) [reply pkg/runtime/bulk_subscriber.go]: Done.

**Comment by DeepanshuA** [comment]
> > Please replicate the code used for custom attributes (cloud event extensions) for the regular gRPC publish path. This was just merged #5516.
> > The entire cloud event represents a map[string]interface{}. All properties minus the well known cloud event fields (id, source, type, spec version, data content type, data field / data field base64) should be removed from the cloud event, and the remainder should be passed through as a struct on the proto.
> 
> I have incorporated suggested change, please re-review @berndverst @mukundansundar

QQ: I realized while testing that now we are sending all fields that come to us in cloudEvent. Earlier only required fields were sent but now any/all OTHER fields are being part of Extensions. Do you foresee any requirement where-in subscriber App would need these extensions? Or we are just sending even un-required data?

**Comment by DeepanshuA** [comment]
I don't what is the logic behind re-requesting review on GH. When I re-request review from someone, it is automatically removing review request from another person (rofl). Quite weird but it happened earlier as well.
So, re-requesting review here - hopefully build and E2E should pass.
@berndverst  @mukundansundar

File: pkg/runtime/bulk_subscriber.go [hunk 3]
```diff
@@ -442,20 +443,77 @@ func (a *DaprRuntime) publishBulkMessageHTTP(ctx context.Context, msg *pubsubBul
...
-		item := &runtimev1pb.TopicEventBulkRequestEntry{
-			EntryId:     entry.EntryId,
-			Event:       entry.Event,
-			ContentType: entry.ContentType,
-			Metadata:    entry.Metadata,
+		item, err := fetchEntry(rawPayload, entry, msg.cloudEvents[i])
+		if err != nil {
+			bulkSubDiag.statusWiseDiag[string(pubsub.Retry)]++
+			setBulkResponseEntry(bulkResponses, i, entry.EntryId, err)
+			continue
```
Comment (mukundansundar) [inline pkg/runtime/bulk_subscriber.go:512]: can you add a test for this scenario?

Comment (DeepanshuA) [reply pkg/runtime/bulk_subscriber.go:512]: added

**Review by mukundansundar — APPROVED** [review]

File: pkg/runtime/bulk_subscriber.go [hunk 4]
```diff
@@ -171,7 +171,7 @@ func (a *DaprRuntime) bulkSubscribeTopic(ctx context.Context, policy resiliency.
 		}
 		var overallInvokeErr error
 		for path, psm := range routePathBulkMessageMap {
-			invokeErr := a.createEnvelopeAndInvokeSubscriber(ctx, psm, topic, psName, msg, route, &bulkResponses, &entryIdIndexMap, path, policy, &bulkSubDiag)
+			invokeErr := a.createEnvelopeAndInvokeSubscriber(ctx, psm, topic, psName, msg, route, &bulkResponses, &entryIdIndexMap, path, policy, &bulkSubDiag, rawPayload)
```
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply pkg/runtime/bulk_subscriber.go:174]: You are correct. Sure, I will make sure that I have another PR (probably next week) taking care of this and if possible, few other functions as well. But, would prefer to keep this PR separate from those changes. Hope that is fine.

File: dapr/proto/runtime/v1/appcallback.proto [hunk 5]
```diff
@@ -118,19 +118,46 @@ message TopicEventResponse {
...
+}
+
 // TopicEventBulkRequestEntry represents a single message inside a bulk request
 message TopicEventBulkRequestEntry {
   // Unique identifier for the message.
   string entry_id = 1;
 
   // The content of the event.
-  bytes event = 2;
+  oneof event {
```
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply dapr/proto/runtime/v1/appcallback.proto:151]: Yeah, had a meeting with maintainers and this was the decision that was made. @artursouza @yaron2

File: dapr/proto/runtime/v1/appcallback.proto [hunk 6]
```diff
@@ -118,19 +118,46 @@ message TopicEventResponse {
...
+    bytes bytes = 2;
+    TopicEventCERequest cloud_event = 3;
+  }
   
   // content type of the event contained.
-  string content_type = 3;
+  string content_type = 4;
```
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply dapr/proto/runtime/v1/appcallback.proto:157]: Sure, I will change that.

File: pkg/runtime/runtime.go [hunk 7]
```diff
@@ -2273,6 +2257,29 @@ func (a *DaprRuntime) publishMessageGRPC(ctx context.Context, msg *pubsubSubscri
...
+	// Create copy of the cloud event with duplicated data removed
+
+	extensions := map[string]interface{}{}
+	for key, value := range cloudEvent {
+		if !cloudEventDuplicateKeys.Has(key) {
+			extensions[key] = value
+		}
+	}
+	extensionsStruct := structpb.Struct{}
+	extensionBytes, jsonMarshalErr := json.Marshal(extensions)
```
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply pkg/runtime/runtime.go:2271]: Well, basically I have just taken out work done in https://github.com/dapr/dapr/pull/5516/files in a separate function. I will see if I can do something better over there. Does it need to be part of this PR though?

File: dapr/proto/runtime/v1/appcallback.proto (hunk 6 above)
[the remark this answers is not in the document]
Comment (mukundansundar) [reply dapr/proto/runtime/v1/appcallback.proto:157]: @ItalyPaleAle I think this proto is specific for bulk subscribe? And it has not been released yet? @DeepanshuA correct me if I am wrong

File: dapr/proto/runtime/v1/appcallback.proto (hunk 5 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply dapr/proto/runtime/v1/appcallback.proto:151]: Alright!

File: dapr/proto/runtime/v1/appcallback.proto (hunk 6 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply dapr/proto/runtime/v1/appcallback.proto:157]: Oh yes, if it's not been released yet (and not implemented by any SDK) then that's fine. But if SDKs have implemented it already, I'd recommend not altering tags.

File: pkg/runtime/bulk_subscriber.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/runtime/bulk_subscriber.go:174]: For sure.

File: pkg/runtime/runtime.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/runtime/runtime.go:2271]: I'm ok with a follow-up PR, but please don't forget about this :)

File: dapr/proto/runtime/v1/appcallback.proto (hunk 6 above)
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply dapr/proto/runtime/v1/appcallback.proto:157]: Yeah, that is not yet released. I assumed you are saying for someone using Nightly Build sort of thing.

**Pull request merged**